### PR TITLE
fix(types): correct inferring when Method has Endponts as Union

### LINKED ISSUE #2607: Wrong type with InferResponseType<> and status code when using .route() method
### What version of Hono are you using?

4.3.1

### What runtime/platform is your app running on?

Bun

### What steps can reproduce the bug?

This code reproduce the bug. You need to create a now router and import it to the current `app` router exposed to the client

```ts
import { Hono } from "hono"
import { hc } from "hono/client"

const myRouter = new Hono()
  .get("/", (c) => {
    const randomBoolean = Math.random() < 0.5
    if (!randomBoolean) {
      return c.json({ error: "not found" }, 404)
    }

    return c.json({ success: true }, 200)
  })

const app = new Hono()
  .basePath("/api")
  .route("/hello", myRouter)

const client = hc<typeof app>("/")

type ResponseType200 = InferResponseType<typeof client.api.hello.$get, 200>
type ResponseType404 = InferResponseType<typeof client.api.hello.$get, 404>
```

### What is the expected behavior?

In the previous code, `ResponseType200` should be of type `{ success: boolean }` and `ResponseType404` should be of type `{ error: string }`

### What do you see instead?

`ResponseType200` and `ResponseType404` are both typed as 
```ts
{ success: boolean } | { error: string }
```

### Additional information

In the [4.3.0 releaseStatus code type]( https://github.com/honojs/hono/releases/tag/v4.3.0) the example show a working example that doesn't use `.route()` method. I think the issue came from the type merging with this method.

## The error
CleanShot 2024-05-04 at 15 38 34@2x

## ✅ Types in the custom router that I want
CleanShot 2024-05-04 at 15 39 47@2x

## ❌ Types of the main router with my custom router imported with the `.route()`
CleanShot 2024-05-04 at 15 40 16@2x

### PR DESCRIPTION
**Pull request by yusukebe**
Fixes #2607

### Author should do the followings, if applicable

- [x] Add tests
- [x] Run tests
- [x] `bun denoify` to generate files for Deno

### CODE REVIEW

**Comment by yusukebe** [comment]
Hi @NamesMT 

Could you review this?

**Comment by armandsalle** [comment]
Oh you are too fast, [I made a PR](https://github.com/honojs/hono/pull/2611) to close [my issue ](https://github.com/honojs/hono/issues/2607) 🙏

Do you have time to explain what your approach does compared to mine? I've i struggled a lot but find a way to fix it!

Thank for your time and your effort ❤️

**Review by armandsalle — APPROVED** [review]
LGTM

**Comment by NamesMT** [comment]
LGTM
Typescript is fun (XD)

File: deno_dist/types.ts [hunk 1]
```diff
@@ -1658,44 +1658,50 @@ type ExtractParams<Path extends string> = string extends Path
...
-            : Input & {
-                // Maps extracted keys, stripping braces, to a string-typed record.
+  [P in keyof OrigSchema as MergePath<SubPath, P & string>]: OrigSchema[P] extends {
+    [M in keyof OrigSchema[P]]: OrigSchema[P][M]
+  }
+    ? {
+        [M in keyof OrigSchema[P]]: MergeEndpointParamsWithPath<OrigSchema[P][M], SubPath>
+      }
+    : never
+}>
```
Comment (NamesMT) [inline deno_dist/types.ts:1668]: Could we make this part and just put
```
Prettify<{
  [P in keyof OrigSchema as MergePath<SubPath, P & string>]: {
    [M in keyof OrigSchema[P]]: MergeEndpointParamsWithPath<OrigSchema[P][M], SubPath>
  }
}>
```

What do you think?

**Review by NamesMT — APPROVED** [review]
LGTM!!

**Comment by NamesMT** [comment]
> Do you have time to explain what your approach does compared to mine? I've i struggled a lot but find a way to fix it!

There was nothing wrong with the old code, it's just TypeScript trying to be smart and merge the types for us, by simply copying the whole block to a new type, we force TS to make it union at that level.

image
=>
image

**Comment by armandsalle** [comment]
> > Do you have time to explain what your approach does compared to mine? I've i struggled a lot but find a way to fix it!
> 
> There was nothing wrong with the old code, it's just TypeScript trying to be smart and merge the types for us, by simply copying the whole block to a new type, we force TS to make it union at that level.
> 
> image => image

TS for libraries is pretty hard 🥲 thank's for the explanations

File: deno_dist/types.ts (hunk 1 above)
Comment (yusukebe) [reply deno_dist/types.ts:1664]: You are right! I've fixed it.

**Comment by yusukebe** [comment]
@NamesMT Thanks!

@armandsalle Yes. TypeScript is hard!

I'll merge this now.

**Pull request merged**